pkg/pgmodel/pgx.go

```diff
@@ -643,37 +643,104 @@ func runCopyFrom(conn pgxConn, in chan copyRequest) {
 		if !ok {
 			return
 		}
-		_, err := conn.CopyFrom(
-			context.Background(),
-			pgx.Identifier{dataSchema, req.table},
-			copyColumns,
-			&req.data.batch,
-		)
+		err := doCopyFrom(conn, req)
 		if err != nil {
-			if pgErr, ok := err.(*pgconn.PgError); ok && strings.Contains(pgErr.Message, "insert/update/delete not permitted") {
-				/* If the error was that the table is already compressed, decompress and try again. */
-				decompressErr := decompressChunks(conn, req.data, req.table)
-				if decompressErr != nil {
-					req.data.reportResults(err)
-					pendingBuffers.Put(req.data)
-					continue
-				}
-
-				req.data.batch.ResetPosition()
-				_, err = conn.CopyFrom(
-					context.Background(),
-					pgx.Identifier{dataSchema, req.table},
-					copyColumns,
-					&req.data.batch,
-				)
-			}
+			err = copyFromErrorFallback(conn, req, err)
 		}
 
 		req.data.reportResults(err)
 		pendingBuffers.Put(req.data)
 	}
 }
 
+// certain errors are recoverable, handle those we can
+//   1. if the data contains duplicates, switch to INSERT ... ON CONFLICT DO NOTHING
+//      so the duplicates are dropped and the other data can go in
+//   2. if the table is compressed, decompress and retry the insertion
+func copyFromErrorFallback(conn pgxConn, req copyRequest, err error) error {
+	// we use either COPY FROM or INSERT depending on if the dataset has duplicates
+	// if we do need to use INSERT, we need to keep doing so, otherwise we'll
+	// fail due to duplicates.
+	insertFn := doCopyFrom
+	for i := 0; err != nil && i < 10; i++ {
+		insertFn, err = tryRecovery(conn, req, insertFn, err)
+		if insertFn == nil || err != nil {
+			return err
+		}
+		err = insertFn(conn, req)
+	}
+	if err != nil {
+		log.Warn("msg", fmt.Sprintf("time out while processing error for %s", req.table), "error", err.Error())
+	}
+	return err
+}
+
+type insertionFunction = func(conn pgxConn, req copyRequest) error
+
+// we can currently recover from two error:
+// If we inserted duplicate data we switch to using INSERT ... ON CONFLICT DO NOTHING
+// to skip redundant data points and try again.
+// If we inserted into a compressed chunk, we decompress the chunk and try again.
+// Since a single batch can have both errors, we need to remember the insert method
+// we're using, so that we deduplicate if needed.
+func tryRecovery(conn pgxConn, req copyRequest, insertFn insertionFunction, err error) (insertionFunction, error) {
+
+	// we only recover from postgres errors right now
+	pgErr, ok := err.(*pgconn.PgError)
+	if !ok {
+		log.Warn("msg", fmt.Sprintf("unexpected error while inserting to %s", req.table), "error", err.Error())
+		return nil, err
+	}
+
+	// if we have duplicate data points try again using
+	// INSERT ... ON CONFLICT DO NOTHING to insert only the new ones
+	if pgErr.Code == "23505" {
+		log.Warn("msg", fmt.Sprintf("duplicate data points inserted to %s", req.table), "table", req.table)
+		req.data.batch.ResetPosition()
+		return doInsert, nil
```

Comment:
since this function internally is aware of the `doInsert` function, does it make sense to drop the insertFn argument, and just return `doCopyFrom` (as doInsert is returned)

Reply:
the issue with that is an sequence of inserts which fails like
```
duplicate key
compressed data
```
after we decompress the data, we still need to `doInsert`, otherwise we'll fail again. (technically retry will catch that, but it'll be yet another round trip) (it's not the best that we have to restart the entire thing as it is, but that's the cost of large batches)